File: mobile_sdk/lib/src/queue/command_queue.dart
import 'dart:collection';
import 'dart:typed_data';

import 'package:robot_protocol/robot_protocol.dart';

class QueuedCommand {
  const QueuedCommand({
    required this.seq,
    required this.command,
    required this.frameBytes,
    required this.isMove,
    this.retries = 0,
    this.lastSentAt,
    this.superseded = false,
  });

  final int seq;
  final RobotCommand command;
  final Uint8List frameBytes;
  final bool isMove;
  final int retries;
  final DateTime? lastSentAt;
  final bool superseded;

  QueuedCommand copyWith({
    int? retries,
    DateTime? lastSentAt,
    bool? superseded,
    bool clearLastSentAt = false,
  }) {
    return QueuedCommand(
      seq: seq,
      command: command,
      frameBytes: frameBytes,
      isMove: isMove,
      retries: retries ?? this.retries,
      lastSentAt: clearLastSentAt ? null : (lastSentAt ?? this.lastSentAt),
      superseded: superseded ?? this.superseded,
    );
  }

  QueuedCommand markSent(DateTime sentAt) => copyWith(lastSentAt: sentAt);

  QueuedCommand bumpRetry(DateTime sentAt) =>
      copyWith(retries: retries + 1, lastSentAt: sentAt);
}

class CommandQueue {
  final Queue<QueuedCommand> _discreteQueue = Queue<QueuedCommand>();
  QueuedCommand? _latestSlot;
  QueuedCommand? _inflight;

  QueuedCommand? get inflight => _inflight;

  bool get hasPending =>
      _inflight != null || _latestSlot != null || _discreteQueue.isNotEmpty;

  void enqueue(QueuedCommand command) {
    if (command.isMove) {
      _latestSlot = command;
      return;
    }
    _latestSlot = null;
    _discreteQueue.add(command);
  }

  void enqueueLatest(QueuedCommand command) {
    _discreteQueue.clear();
    _inflight = _inflight?.copyWith(superseded: true);
    _latestSlot = command;
  }

  QueuedCommand? promoteNext(DateTime sentAt) {
    if (_inflight != null) {
      return null;
    }

    final next =
        _discreteQueue.isNotEmpty ? _discreteQueue.removeFirst() : _latestSlot;
    if (next == null) {
      return null;
    }

    if (identical(next, _latestSlot)) {
      _latestSlot = null;
    }

    _inflight = next.markSent(sentAt);
    return _inflight;
  }

  bool acknowledge(int seq) {
    if (_inflight == null || _inflight!.seq != seq) {
      return false;
    }
    _inflight = null;
    return true;
  }

  QueuedCommand? retryCurrent(DateTime sentAt, int maxRetries) {
    final current = _inflight;
    if (current == null) {
      return null;
    }
    if (current.retries >= maxRetries) {
      _inflight = null;
      return null;
    }
    _inflight = current.bumpRetry(sentAt);
    return _inflight;
  }

  void resetInflightTiming() {
    final current = _inflight;
    if (current == null) {
      return;
    }
    _inflight = current.copyWith(clearLastSentAt: true);
  }

  void dropCurrent() {
    _inflight = null;
  }
}

File: robot_server/robot_server/runtime/command_queue.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, replace
from typing import Deque, Optional


@dataclass(frozen=True)
class QueuedCommand:
    seq: int
    frame: bytes
    is_move: bool
    retries: int = 0
    last_sent_at: Optional[float] = None

    def mark_sent(self, sent_at: float) -> "QueuedCommand":
        return replace(self, last_sent_at=sent_at)

    def bump_retry(self, sent_at: float) -> "QueuedCommand":
        return replace(self, retries=self.retries + 1, last_sent_at=sent_at)


class CommandQueue:
    def __init__(self) -> None:
        self._discrete_queue: Deque[QueuedCommand] = deque()
        self._move_slot: Optional[QueuedCommand] = None
        self._inflight: Optional[QueuedCommand] = None

    @property
    def inflight(self) -> Optional[QueuedCommand]:
        return self._inflight

    @property
    def has_pending(self) -> bool:
        return self._inflight is not None or self._move_slot is not None or bool(self._discrete_queue)

    def enqueue(self, command: QueuedCommand) -> None:
        if command.is_move:
            self._move_slot = command
            return
        self._discrete_queue.append(command)

    def promote_next(self, sent_at: float) -> Optional[QueuedCommand]:
        if self._inflight is not None:
            return None

        next_command = self._discrete_queue.popleft() if self._discrete_queue else self._move_slot
        if next_command is None:
            return None

        if next_command is self._move_slot:
            self._move_slot = None

        self._inflight = next_command.mark_sent(sent_at)
        return self._inflight

    def ack(self, seq: int) -> bool:
        if self._inflight is None or self._inflight.seq != seq:
            return False
        self._inflight = None
        return True

    def retry_current(self, sent_at: float, max_retries: int) -> Optional[QueuedCommand]:
        if self._inflight is None:
            return None
        if self._inflight.retries >= max_retries:
            self._inflight = None
            return None
        self._inflight = self._inflight.bump_retry(sent_at)
        return self._inflight

    def drop_current(self) -> None:
        self._inflight = None
